app/services/work_package_bulk_update_service.rb

```diff
@@ -0,0 +1,156 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+class WorkPackageBulkUpdateService
+  attr_accessor :projects,
+                :work_packages
+
+  def initialize(work_packages, projects)
+    @projects      = projects
+    @work_packages = work_packages
+    @project       = @projects.first if @projects.size == 1
+  end
+
+  def run(params)
+    prepare(params)
+  end
+
+  def available_attributes
+    {
+      available_statuses: @available_statuses,
+      assignables:        @assignables,
+      responsibles:       @responsibles,
+      types:              @types,
+      custom_fields:      @custom_fields,
+      target_project:     @target_project,
+      allowed_projects:   @allowed_projects,
+      copy:               @copy,
+      notes:              @notes
+    }
+  end
+
+  def save(params, copy = false)
+    prepare(params)
+    unsaved_work_package_ids = []
+    moved_work_packages      = []
+    @work_packages.sort!
+
+    @work_packages.each do |work_package|
+      work_package.reload
+
+      if copy
+        Redmine::Hook.call_hook(:controller_work_packages_move_before_save,
+                                { params:         params,
+                                  work_package:   work_package,
+                                  target_project: @target_project,
+                                  copy:           !!@copy })
+
+        if r = work_package.move_to_project(@target_project,
+                                            @new_type, { copy:         @copy,
+                                                         attributes:   permitted_params(params),
+                                                         journal_note: @notes })
+          moved_work_packages << r
+        else
+          unsaved_work_package_ids << work_package.id
+        end
+
+      else
+        work_package.add_journal(User.current, @notes)
+
+        # filter parameters by whitelist and add defaults
+        attributes = parse_params_for_bulk_work_package_attributes params, work_package.project
+        work_package.assign_attributes attributes
+
+        Redmine::Hook.call_hook(:controller_work_package_bulk_before_save,
+                                params: params, work_package: work_package)
+        JournalObserver.instance.send_notification = params[:send_notification] == '0' ? false : true
+        unless work_package.save
+          unsaved_work_package_ids << work_package.id
+        end
+
+      end
+    end
+
+    { moved_work_packages:      moved_work_packages,
+      unsaved_work_package_ids: unsaved_work_package_ids,
+      copy: @copy }
+  end
+
+  private
+
+  def prepare(params)
+    @work_packages.sort!
+    @copy             = params.has_key? :copy
```

Comment:
Feedback from my side is coming late now, but this copy-parameter looks kind of messy.

The method `save` has a parameter called `params` and one called `copy`, the former passed into this method and checking whether it contains a key called `:copy`.

This is the first red flag for me, because the `params` hash may contain a key called `copy` as well as there may be the copy parameter.

The second red flag is that it suffices that the `:copy` key is present for `@copy` to be true. That means `{ copy: false }` would make `@copy` true.